client/src/util/centroid.js

```diff
@@ -6,100 +6,131 @@ import { unassignedCategoryLabel } from "../globals";
+    // Fetch the categorical value of the current cell
     const categoryValue = categoryArray[i];
 
     // Get the index of the categoryValue within the category
-    // If the category is truncated and this value is removed,
-    // it will not be assigned a category value and will not be
-    // labeled on the graph
-    const categoryValueIndex = categoricalSelection[
-      categoryName
-    ].categoryValueIndices.get(categoryValue);
-
-    // Check to see if the current category is a user created annotation
-    // if the user created this category, do not create a label for the `unassigned` value
-    const isUserAnno = schemaObsByName[categoryName].writable;
+    const categoryValueIndex = categoryValueIndices.get(categoryValue);
 
+    // If the category is truncated and this value is removed,
+    //  it will not be assigned a category value and will not be
+    //  labeled on the graph
+    // If the user created this category,
+    //  do not create a label for the `unassigned` value
     if (
       categoryValueIndex !== undefined &&
       !(isUserAnno && categoryValue === unassignedCategoryLabel)
     ) {
-      // Get the number of cells which are in the category value
-      const numInCategoryValue =
-        categoricalSelection[categoryName].categoryValueCounts[
-          categoryValueIndex
-        ];
-
-      // Create/fetch the valueArray,
-      // which is what the key points to in the `coordinates` hashmap
-      const valueArray = coordinates.get(categoryValue) || [
-        false, // hasFinite
-        0, // index
-        new Float32Array(numInCategoryValue), // x coordinates
-        new Float32Array(numInCategoryValue) // y coordinates
-      ];
-      const index = valueArray[1];
-      let hasFinite = valueArray[0];
-
-      hasFinite =
-        hasFinite ||
+      // Create/fetch the scratchpad value
+      let value = scratchPad.get(categoryValue);
+      if (value === undefined) {
+        // Get the number of cells which are in the categorical value
+        const numInCategoricalValue = categoryValueCounts[categoryValueIndex];
+        value = {
+          hasFinite: false,
+          xCoordinates: new Float32Array(numInCategoricalValue),
+          yCoordinates: new Float32Array(numInCategoricalValue),
+          length: 0
+        };
+      }
+
+      value.hasFinite =
+        value.hasFinite ||
         (Number.isFinite(layoutXArray[i]) && Number.isFinite(layoutYArray[i]));
 
-      valueArray[0] = hasFinite;
-      valueArray[1] = index + 1;
-      valueArray[2][index] = layoutXArray[i];
-      valueArray[3][index] = layoutYArray[i];
+      const coordinatesLength = value.length;
 
-      coordinates.set(categoryValue, valueArray);
+      value.xCoordinates[coordinatesLength] = layoutXArray[i];
+      value.yCoordinates[coordinatesLength] = layoutYArray[i];
+
+      value.length = coordinatesLength + 1;
+
+      scratchPad.set(categoryValue, value);
```

Comment:
why do the redundant set() on each value?   If the `value` object is already in the Map(), you don't need to add it again.

Suggestion:
* when you first initialize the `value` object, at line 62, perform the set().
* remove the set() on each iteration through the loop.

in other words,
* you should do a set once per category label value
* you should do a get on each iteration.

Reply:
I'm a little confused here, if there value object is being mutated and then we are inserting the updated objected via `set()`  is that redundant? Since the object returned by `get()` is a copy, any mutations will be lost once we exit scope if we don't insert.

Reply:
get does not return a copy - it returns the actual object

Reply:
You're right, when I tested it earlier I was just using primitives as the values in the hashmap and wasn't getting the expected values *facepalm*

Reply:
fix commited